### PR TITLE
Initial implementation of `stty`

### PR DESCRIPTION
**Pull request by tertsdiepraam**
I got annoyed that there was no `stty` yet, so I started working on it :)

In this initial version:
- ~Only binary flags are implemented (i.e. attributes that are enabled and disabled). This means that attributes like `tabN` where `N` is a number between `0` and `3` is missing.~ This is now supported.
- `--all` is implemented.
- `-F/--file` and `-g/--save` are ignored.
- Special characters, special settings and combined settings are missing.
- The `nix` crate is used (not `libc`). Sadly, here are already some attributes that seem to be missing in `nix`.

The tests are broken on my machine with the error:
```
thread 'main' panicked at 'Could not get terminal attributes: ENOTTY', src/uu/stty/src/stty.rs:83:44
```

So the stdout in `cargo test` does not seem to be a tty with terminal attributes. This might mean that we can't test `stty` at all? I'd like to try to find a workaround though.

### CODE REVIEW

**Comment by sylvestre** [comment]
wahou, amazing :)

**Comment by anastygnome** [comment]
@tertsdiepraam can we use crossterm for this (or are we using it ?) 

Maybe we will have to use shell scripts for that TTY issue.

https://docs.rs/crossterm/latest/crossterm/index.html

**Comment by tertsdiepraam** [comment]
I'm not currently using `crossterm`, but `nix`, which has support for most of the flags that we need for this. `crossterm` seems to be more about writing to the terminal than changing it's characteristics. The terminal size could come in handy but `crossterm` doesn't seem to support setting the size. So I'm afraid it will have to stay as low level as it is.

**Comment by anastygnome** [comment]
@tertsdiepraam I thought so as well, but maybe it could be an inspiration for which calls we should use.

File: src/uu/stty/src/stty.rs [hunk 1]
```diff
@@ -120,7 +120,15 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
 fn stty(opts: &Options) -> UResult<()> {
     // TODO: Figure out the right error message
     let mut termios = tcgetattr(opts.file).expect("Could not get terminal attributes");
+
     if let Some(settings) = &opts.settings {
+        if opts.save || opts.all {
```
Comment (anastygnome) [inline src/uu/stty/src/stty.rs:125]: should be and and right ?? 
This is an or operator.

Comment (tertsdiepraam) [reply src/uu/stty/src/stty.rs:125]: Nope, the `||` is correct here I think. The error is about combining the `--save` flag or the `--all` with settings that you want to apply. Basically, `stty` will refuse to both print and change settings. I.e.:
```
❯ stty --save nl0
stty: when specifying an output style, modes may not be set 
❯ stty --all nl0
stty: when specifying an output style, modes may not be set
```
I did just realize that I can add a test for this if I restructure a bit, so I'm gonna do that.

Comment (tertsdiepraam) [reply src/uu/stty/src/stty.rs:125]: Oh but there is also a separate error for using both `--save` and `--all`, which I should add:
```
❯ stty --all --save
stty: the options for verbose and stty-readable output styles are
mutually exclusive
```

**Comment by tertsdiepraam** [comment]
I've ignored the tests that need to be run from a `tty` for now, but added some for the argument parsing. I've also added the special settings for the terminal (baudrate, columns, rows and line), but they can only be printed for now (not changed). With those changes this is at least ready to be considered somewhat of a viable first version, though of course there is much to be done :)

File: src/uu/stty/src/stty.rs (hunk 1 above)
Comment (anastygnome) [reply src/uu/stty/src/stty.rs]: Yes, the comment reminded me of that ^^'

**Comment by tertsdiepraam** [comment]
Apparently it needs some work on mac. It will take me a while to fix it though. If someone else wants to fix it up for mac, feel free to do so in the meantime.

**Comment by anastygnome** [comment]
@tertsdiepraam I belive mac os has no support for the cline field in termios. 

Hence, should we ignore it ?

**Comment by tertsdiepraam** [comment]
Yeah I think that makes sense. The other issue with mac is that the baudrate is represented with a u32 instead of the baudrate enum.

File: src/uu/stty/src/stty.rs [hunk 2]
```diff
@@ -141,16 +174,42 @@ fn stty(opts: &Options) -> UResult<()> {
...
+        print_settings(&termios, opts).expect("TODO: make proper error here from nix error");
     }
     Ok(())
 }
 
-fn print_settings(termios: &Termios, opts: &Options) {
+fn print_terminal_size(termios: &Termios, opts: &Options) -> nix::Result<()> {
+    let speed = cfgetospeed(termios);
+    for (text, baud_rate) in BAUD_RATES {
+        if *baud_rate == speed {
```
Comment (anastygnome) [inline src/uu/stty/src/stty.rs:181]: ```rust if *baud_rate as u32 == speed as u32```
seems to work on my system and should fix this issue.

File: src/uu/stty/src/flags.rs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+use crate::Flag;
+use nix::sys::termios::{
+    BaudRate, ControlFlags as C, InputFlags as I, LocalFlags as L, OutputFlags as O,
+};
+
+pub const CONTROL_FLAGS: [Flag<C>; 12] = [
+    Flag::new("parenb", C::PARENB),
+    Flag::new("parodd", C::PARODD),
+    Flag::new("cmspar", C::CMSPAR),
```
Comment (sylvestre) [inline src/uu/stty/src/flags.rs:19]: I guess you saw this error?

Comment (sylvestre) [reply src/uu/stty/src/flags.rs:19]: @tertsdiepraam ^^(in case you missed it :)

Comment (tertsdiepraam) [reply src/uu/stty/src/flags.rs:19]: I did see it, but I'm on holiday at the moment so I'm unable to fix it 😄 I'll take a look wt it in a few weeks

**Comment by tertsdiepraam** [comment]
Okay, so I think it's correct now. I can't test with mac, but I get the correct speed on FreeBSD. There are of course some minor issues and missing functionality with this implementation, but I think we can merge this and I can start opening issues for those.

**Comment by sylvestre** [comment]
sounds good, let's wait for the CI :)
I can do more testing on mac if you need

**Comment by tertsdiepraam** [comment]
I'm not sure what's going on with the `Cargo.lock` format. I can't find anything important that has changes with respect to main.

**Comment by sylvestre** [comment]
Bravo !
would you like to do a release now? :)

**Comment by tertsdiepraam** [comment]
Yes, I'll start working on it!

**Pull request merged**